### PR TITLE
Support `mir_alloc`, `mir_alloc_mut`, and `mir_points_to`

### PR DESCRIPTION
**Pull request by RyanGlScott**
This patch adds support for specifying the memory layouts of allocations in MIR specifications through the following commands:

* `mir_alloc`, which allocates an immutable reference, and `mir_alloc_mut`, which allocates a mutable reference.
* `mir_points_to`, which declares that a reference points to a given value in a pre- or post-condition.

These commands behave much like `llvm_alloc` and `llvm_points_to`, and the code for the MIR commands looks quite similar to the LLVM code as a result. The largest difference implementation-wise is that I need to call into the `crucible-mir` memory model in order to read values from and write values to references. This patch therefore bumps the `crucible` submodule to bring in the changes from https://github.com/GaloisInc/crucible/issues/1105, which make it much more straightforward to use the `crucible-mir` memory model in a SAW setting.

Along the way, this patch also performs some mild refactoring:

* This adds a `ConditionMetadata` field to `MirPointsTo` for the benefit of using it in `learnPointsTo`. This required moving around some `ConditionMetadata` values in `crucible-mir-comp` as a result.
* This moves the `CheckPointsToType` data type from `SAWScript.Crucible.LLVM.Builtins` to `SAWScript.Crucible.Common.Setup.Builtins`. This is because `CheckPointsToType` is now used in the `saw-remote-api` code that powers `mir_points_to` (see `SAWServer.MIRCrucibleSetup`), and it doesn't make much sense for this code to import LLVM-specific code.

This checks off one box in #1859.

### CODE REVIEW

File: src/SAWScript/Crucible/MIR/Builtins.hs [hunk 1]
```diff
@@ -252,6 +254,61 @@ mir_postcond term = MIRSetupM $ do
...
+  Setup.CrucibleSetupT MIR TopLevel Mir.Ty
+mir_points_to_check_lhs_validity ref loc =
+  do cc <- getMIRCrucibleContext
+     st <- get
+     let env = MS.csAllocations (st ^. Setup.csMethodSpec)
+         nameEnv = MS.csTypeNames (st ^. Setup.csMethodSpec)
+     refTy <- typeOfSetupValue cc env nameEnv ref
+     case refTy of
+       Mir.TyRef referentTy _ -> pure referentTy
+       _ -> throwCrucibleSetup loc $ "lhs not a reference type: " ++ show refTy
```
Comment (RyanGlScott) [inline src/SAWScript/Crucible/MIR/Builtins.hs:310]: This code checks if the reference passed to `mir_points_to` is actually a reference type.

File: src/SAWScript/Crucible/MIR/Builtins.hs [hunk 2]
```diff
@@ -476,11 +533,34 @@ setupPrePointsTos ::
...
+        Refl <-
+          case W4.testEquality (mp^.mpType) (shapeType referentTy) of
+            Just r -> pure r
+            Nothing ->
+              panic "setupPrePointsTos"
+                    [ "Unexpected type mismatch between reference and referent"
+                    , "Reference type: " ++ show (mp^.mpType)
+                    , "Referent type:  " ++ show (shapeType referentTy)
+                    ]
+        Mir.writeMirRefIO bak globals Mir.mirIntrinsicTypes (mp^.mpRef) referentVal
```
Comment (RyanGlScott) [inline src/SAWScript/Crucible/MIR/Builtins.hs:563]: This is one of the places in the patch that involves the `crucible-mir` memory model. Here, we write the right-hand side values for each `mir_points_to` statement in a specification's preconditions.

File: src/SAWScript/Crucible/MIR/Builtins.hs [hunk 3]
```diff
@@ -1044,6 +1122,41 @@ cryptolTypeOfActual mty =
...
+
+     ptr <- liftIO $ do
+       zero <- W4.bvLit sym W4.knownRepr $ BV.mkBV W4.knownRepr 0
+       Mir.subindexMirRefIO bak iTypes tpr ref zero
+     pure $ Some MirPointer
+       { _mpType = tpr
+       , _mpMutbl = ma^.maMutbl
+       , _mpMirType = ma^.maMirType
+       , _mpRef = ptr
+       }
```
Comment (RyanGlScott) [inline src/SAWScript/Crucible/MIR/Builtins.hs:1158]: This is where most of the magic powering `mir_alloc{,_mut}` happens. Using the `crucible-mir` memory model, this allocates a fresh reference, writes uninitialized memory to it, and returns the resulting pointer. This code was largely inspired by [similar code in `crucible-mir`](https://github.com/GaloisInc/crucible/blob/c19fc2c53851cdd162ae1ccca14032467e3c2fd1/crucible-mir/src/Mir/TransCustom.hs#L1393-L1402).

File: src/SAWScript/Crucible/MIR/MethodSpecIR.hs [hunk 4]
```diff
@@ -104,12 +104,17 @@ mccSym = to (\mcc -> mccWithBackend mcc backendGetSym)
...
 type instance MS.Pointer' MIR sym = Some (MirPointer sym)
 
-
-data MirPointsTo = MirPointsTo MS.AllocIndex [MS.SetupValue MIR]
+-- | Unlike @LLVMPointsTo@ and @JVMPointsTo@, 'MirPointsTo' contains a /list/ of
+-- 'MS.SetupValue's on the right-hand side. This is due to how slices are
+-- represented in @crucible-mir-comp@, which stores the list of values
+-- referenced by the slice. The @mir_points_to@ command, on the other hand,
+-- always creates 'MirPointsTo' values with exactly one value in the list (see
+-- the @firstPointsToReferent@ function in "SAWScript.Crucible.MIR.Override").
```
Comment (RyanGlScott) [inline src/SAWScript/Crucible/MIR/MethodSpecIR.hs:112]: I added this comment here...

File: src/SAWScript/Crucible/MIR/Override.hs [hunk 5]
```diff
@@ -146,6 +147,18 @@ enforceDisjointness cc loc ss =
...
+-- referent on the right-hand side. As a result, this function should never
+-- fail.
+firstPointsToReferent ::
+  MonadFail m => [MS.SetupValue MIR] -> m (MS.SetupValue MIR)
+firstPointsToReferent referents =
+  case referents of
+    [referent] -> pure referent
+    _ -> fail $
+      "Unexpected mir_points_to statement with " ++ show (length referents) ++
+      " referent(s)"
```
Comment (RyanGlScott) [inline src/SAWScript/Crucible/MIR/Override.hs:160]: ...to justify what this code is doing.

File: src/SAWScript/Crucible/MIR/Override.hs [hunk 6]
```diff
@@ -226,8 +239,16 @@ learnPointsTo ::
...
-  panic "learnPointsTo" ["not yet implemented"]
+learnPointsTo opts sc cc spec prepost (MirPointsTo md allocIdx referents) =
+  mccWithBackend cc $ \bak ->
+  do let col = cc ^. mccRustModule . Mir.rmCS ^. Mir.collection
+     globals <- OM (use overrideGlobals)
+     Some mp <- resolveAllocIndexMIR allocIdx
+     let mpMirTy = mp^.mpMirType
+     let mpTpr = tyToShapeEq col mpMirTy (mp^.mpType)
+     val <- firstPointsToReferent referents
+     v <- liftIO $ Mir.readMirRefIO bak globals Mir.mirIntrinsicTypes (mp^.mpType) (mp^.mpRef)
```
Comment (RyanGlScott) [inline src/SAWScript/Crucible/MIR/Override.hs:250]: The final place in the patch that makes use of the `crucible-mir` memory model, this time to read from a reference.

File: src/SAWScript/Crucible/Common/Setup/Builtins.hs [hunk 7]
```diff
@@ -83,3 +84,19 @@ crucible_execute_func args = do
...
+data CheckPointsToType ty
+  = CheckAgainstPointerType
+    -- ^ Check the type of the RHS value against the type that the LHS points to.
+    --   Used by commands such as @llvm_{conditional_}points_to@ and
+    --   @mir_points_to@.
+  | CheckAgainstCastedType ty
+    -- ^ Check the type of the RHS value against the provided @ty@, which
+    --   the LHS pointer is casted to.
+    --   This is currently used by @llvm_{conditional_}points_to_at_type@ only.
+  deriving (Functor, Foldable, Traversable)
```
Comment (RyanGlScott) [inline src/SAWScript/Crucible/Common/Setup/Builtins.hs:102]: Reviewers' note: this isn't new code, but rather code that was moved from `SAWScript.Crucible.LLVM.Builtins`. This data type is now used in the MIR-related code in `saw-remote-api`, so it no longer made sense for it to live in an LLVM-specific module.

**Review by RyanGlScott — COMMENTED** [review]
Like with #1904, much of this code is mechanically adapted from the LLVM/JVM backends. I've made an effort to highlight the interesting bits that are MIR-specific.

**Review by spernsteiner — APPROVED** [review]
This looks reasonable to me, including all the MIR memory model integration.

I was worried that the current approach of extracting MIR types from SetupValues wouldn't be sufficient for handling slices, but that will be addressed by #1914.

**Pull request merged**